featuregates/controller/pkg/feature/feature_suite_test.go

```diff
@@ -0,0 +1,553 @@
+	"k8s.io/apimachinery/pkg/runtime/serializer/yaml"
+	"k8s.io/apimachinery/pkg/types"
+	yamlutil "k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+	"sigs.k8s.io/controller-runtime/pkg/envtest"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+
+	corev1alpha2 "github.com/vmware-tanzu/tanzu-framework/apis/core/v1alpha2"
+)
+
+func TestFeaturegate(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "Featuregate Suite")
+}
+
+var (
+	cfg       *rest.Config
+	k8sClient client.Client
+	testEnv   *envtest.Environment
+	ctx       context.Context
+	cancel    context.CancelFunc
+	setupLog  = ctrl.Log.WithName("controllers").WithName("FeatureGates")
+
+	timeout  = 5 * time.Second
+	interval = 100 * time.Millisecond
+)
+
+func generateCertificateAndManifests() error {
+	cert := &x509.Certificate{
+		SerialNumber: big.NewInt(2019),
+		Subject: pkix.Name{
+			Organization:  []string{"Company, INC."},
+			Country:       []string{"US"},
+			Province:      []string{""},
+			Locality:      []string{"San Francisco"},
+			StreetAddress: []string{"Golden Gate Bridge"},
+			PostalCode:    []string{"94016"},
+		},
+		DNSNames:              []string{"tanzu-featuregates-webhook-service.tkg-system.svc"},
+		NotBefore:             time.Now(),
+		NotAfter:              time.Now().AddDate(10, 0, 0),
+		IsCA:                  false,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
+		KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		BasicConstraintsValid: true,
+	}
+
+	key, err := rsa.GenerateKey(rand.Reader, 4096)
+	if err != nil {
+		return err
+	}
+
+	certBytes, err := x509.CreateCertificate(rand.Reader, cert, cert, &key.PublicKey, key)
+	if err != nil {
+		return err
+	}
+
+	certPem := new(bytes.Buffer)
+	pem.Encode(certPem, &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: certBytes,
+	})
+
+	keyPem := new(bytes.Buffer)
+	pem.Encode(keyPem, &pem.Block{
+		Type:  "RSA PRIVATE KEY",
+		Bytes: x509.MarshalPKCS1PrivateKey(key),
+	})
+
+	caBundleBytes := certPem.Bytes()
+	err = os.WriteFile("/tmp/feature-webhook.crt", caBundleBytes, 0644)
+	if err != nil {
+		return err
+	}
+
+	err = os.WriteFile("/tmp/feature-webhook.key", keyPem.Bytes(), 0644)
+	if err != nil {
+		return err
+	}
+
+	input, err := ioutil.ReadFile("testdata/webhook.yaml")
+	if err != nil {
+		log.Fatalln(err)
+	}
+
+	lines := strings.Split(string(input), "\n")
+
+	for i, line := range lines {
+		if strings.Contains(line, "Cg==") {
+			lines[i] = strings.Replace(lines[i], "Cg==", base64.StdEncoding.EncodeToString(caBundleBytes), 1)
+		}
+	}
+	output := strings.Join(lines, "\n")
+	err = ioutil.WriteFile("/tmp/fearuregate-webhook.yaml", []byte(output), 0644)
+	if err != nil {
+		log.Fatalln(err)
+	}
```

Comment:
I think a cleaner approach would be to craft appropriate Go objects in memory and manipulate them. We are reading and writing objects to a file only to decode them later as unstructured objects for creation.

Reply:
That's correct! I will explore options.

Reply:
Done. Please verify if it makes sense.